MIPRO optimizer updates for paper release

XenonMolecule opened the pull request:
Updates the MIPRO optimizer to add in program aware proposal, minibatching, and refactor to dspy.propose pattern in line with the paper release: https://arxiv.org/abs/2406.11695

Also updates the MIPRO HotPotQA notebook and creates a ScoNe notebook.  Note that neither notebook has a functioning cache yet.

Old MIPRO is preserved temporarily in `mipro_optimizer_deprecated.py`

XenonMolecule commented:
Note that this PR makes **breaking changes** to the MIPRO optimizer and is not backwards compatible.

rawwerks commented:
can't wait! this is going to be so awesome.

mikeedjones commented on dsp/primitives/predict.py:
```diff
@@ -112,19 +77,45 @@ def do_generate(
+            # Recurse with greedy decoding and a shorter length.
+            max_tokens = kwargs.get("max_tokens", dsp.settings.lm.kwargs["max_tokens"])
+            max_tokens = min(max(75, max_tokens // 2), max_tokens)
+            new_kwargs = {
+                **kwargs,
+                "max_tokens": max_tokens,
+                "n": 1,
+                "temperature": 0.0,
+            }
```
setting n=1 here causes a regression on issues https://github.com/stanfordnlp/dspy/issues/914 and https://github.com/stanfordnlp/dspy/issues/749

mikeedjones commented:
This MR includes a reversion of https://github.com/stanfordnlp/dspy/pull/920

XenonMolecule commented:
Hey Mike, thanks for looking at this!  I reverted predict.py back because it was causing a few test programs I ran to crash, not to mention the MIPRO optimizer itself.  This was specifically due to the max_depth assertion.  We’ve been developing off an old branch of the repo so when I brought everything up to date it seemed like an issue in the predictor logic.  I’m happy to remove my changes to predict.py, but do you have suggestions for handling the case where the LM doesn’t generate all fields?  Is it a matter of changing the kwargs to predict to increase the max_depth?

mikeedjones commented:
My guess is the crash would be caused by the LM skipping a field in one of the candidate generations and that skip not being picked up by the extend generation logic. Then, because the LM has moved on to another filed in the signature it would be very unlikely to go back and fill the earlier field, independently of how many levels of recursion.

Happy to do some investigation as to why - the extend generation logic should roll back to the first skipped field so if the above is the behaviour, then it's a bug in #920 's implementation. 

As to why that is never seen in the old logic - I'm stumped. I tried to replicate as closely as possible the logic in there. Something else to figure out!

mikeedjones commented on dspy/propose/utils.py:
```diff
@@ -0,0 +1,200 @@
+    instructions = re.findall(r'"([^"]*)"', instruction_string)
+    return instructions
+
+# def get_program_instruction_set_string(program):
+#     instruction_set_dict = {}
+#     for i, pred in enumerate(program.predictors()):
+#         pred_instructions = get_signature(pred).instructions
+#         instruction_set_dict[f"Module {i} Instruction"] = pred_instructions
+#     return '\n'.join(f"{key}: {value}" for key, value in instruction_set_dict.items())
+
+def get_program_instruction_set_string(program):
+    instruction_list = []
+    for _, pred in enumerate(program.predictors()):
+        pred_instructions = get_signature(pred).instructions
+        instruction_list.append(f"\"{pred_instructions}\"")
+    # Joining the list into a single string that looks like a list
+    return f"[{', '.join(instruction_list)}]"
```
can one be removed?

mikeedjones commented on dspy/teleprompt/hard_bootstrap.py:
```diff
@@ -0,0 +1,239 @@
+                    break
+
+            # hard_examples.append(example)
+        #breakpoint()
```
remove debugging statements?

mikeedjones commented on dspy/teleprompt/hard_bootstrap.py:
```diff
@@ -0,0 +1,239 @@
+                    print("Bootstrapping complete!")
+                    break
+
+            # hard_examples.append(example)
```
Hard examples is never appended to?

mikeedjones commented on dspy/teleprompt/hard_bootstrap.py:
```diff
@@ -0,0 +1,239 @@
+
+        try:
+            with dsp.settings.context(trace=[], **self.teacher_settings):
+                # breakpoint()
```
debugging statements

mikeedjones commented on dspy/teleprompt/hard_bootstrap.py:
```diff
@@ -0,0 +1,239 @@
+                        predictor.demos = [x for x in predictor.demos if x != example]
+
+                    prediction = teacher(**example.inputs())
+                    # print(f"Example: {example}")
+                    # print(f"Prediction: {prediction}")
```
debugging statements

mikeedjones commented on dspy/teleprompt/hard_bootstrap.py:
```diff
@@ -0,0 +1,239 @@
+                success = (self.metric is None) or self.metric(example, prediction, trace)
+
+                lm.kwargs["temperature"] = original_lm_temp
+                # print(success, example, prediction)
```
debugging statements

mikeedjones commented on dspy/teleprompt/hard_bootstrap.py:
```diff
@@ -0,0 +1,239 @@
+    
+    def _bootstrap_one_example(self, example, round_idx=0):
+        name2traces = self.name2traces
+        teacher = self.teacher #.deepcopy()
```
should this be deepcopied or not?

mikeedjones commented on dspy/teleprompt/hard_bootstrap.py:
```diff
@@ -0,0 +1,239 @@
+                    if total_successes > (total_attempts - min_failed_attempts):
+                        break
+                # if this example failed the right number of times and is therefore 'hard', let's add it as one of our demo examples
+                # print(f"Total Successes: {total_successes} | Total Failures: {total_attempts-total_successes}")
```
I thought dspy had moved from print statements to logging?

mikeedjones commented on dspy/teleprompt/hard_bootstrap.py:
```diff
@@ -0,0 +1,239 @@
+        self.trainset = trainset
+        self.valset = valset
+
+        raise Exception("NOTE THAT THIS HAS BEEN EXPENSIVE TO RUN. CHECK IN WITH KRISTAOO BEFORE DOING A FULL RUN.")
```
KRISTA00 is gonna be busy!

mikeedjones commented on dspy/propose/utils.py:
```diff
@@ -0,0 +1,200 @@
+    # Joining all the field strings
+    return '\n'.join(output)
+
+# def create_predictor_level_history_string(base_program, predictor_i, trial_logs, top_n):
```
can this be removed?

mikeedjones commented on dspy/propose/utils.py:
```diff
@@ -0,0 +1,200 @@
+    
+    return instruction_set_history_string
+
+# def parse_list_of_instructions(instruction_string):
```
can this be removed?

XenonMolecule commented:
Thanks Mike I addressed your comments and cleaned up the code from the dev setting based on your comments!  In particular I also removed hard bootstrapping from this PR as we aren't using it for this particular Optimizer.

If you are looking for a specific test setting where the `propose.py` was failing, you can check out [examples/qa/hotpot/hotpotqa_with_MIPRO.ipynb](this notebook).  I'll take a closer look at #1172.

mikeedjones commented:
Thanks! Based on what I saw with the tests I think it should be possible to inspect the cache of that notebook to see if the implemented extend gen logic is exhibiting the issue I saw in #1172. 

That said, the extend generation logic from #920 doesn't pass the tests in #1172 either - and seems to be causing issues like #1108. Would be good to get the view from the author of the extend gen logic and it's expected behaviour. The logic was in the initial commit so probably @okhat?

XenonMolecule commented on dspy/teleprompt/mipro_optimizer_v2.py:
```diff
@@ -0,0 +1,505 @@
+
+* prompt_model: The model used for prompt generation. When unspecified, defaults to the model set in settings (i.e., dspy.settings.configure(lm=task_model)).
+* task_model: The model used for prompt generation. When unspecified, defaults to the model set in settings (i.e., dspy.settings.configure(lm=task_model)).
+* teacher_settings: The settings used for the teacher model. When unspecified, defaults to the settings set in settings (i.e., dspy.settings.configure(lm=task_model)).
```
Edited!  Thanks for the suggestion!

mikeedjones commented:
@XenonMolecule - did you manage to track down whether or not the reverted extend generation logic was leading to the prediction from the language model ending up in the wrong OutputField? Would be great to understand what I missed in #1172 Thank you!
